### PR TITLE
Add GlobalManager system

### LINKED ISSUE #6777: Add a singleton based plugin system to FEProblem
### Description of the enhancement or error report

`FEProblem` is becoming a monolithic hub for everything _global_ in a MOOSE simulation. Every feature that needs some kind of global manger code adds to FEProblem.h and (FEProblem.C). Frequent additions trigger massive recompiles. Adding more and more code makes the class harder and harder to maintain.

Module (and app) developers that want to register global manager code need to make changes in framework. It would be nice if developers could accomplish this within their own codebase.

So here is my latest crackpot idea: **GlobalManagers**. This new _minisystem_ will allow users to register objects in `FEProblem` (and obtain references to them). These objects or _managers_ contain all the code and member variables that were previously tacked onto `FEProblem`. I'm working on a PR where I demonstrate the capability by using it to move all of the `RandomInterface` related code out of FEProblem and into its own manager class. I have plans to add a manager class that provides meta data on derivative parsed materials, which will drastically simplify their use.

More concrete, I propose adding a method `FEProblem::getGlobal<T>()` that returns a reference to a `T` object (which is derived from `GlobalManager`). There is no _name_ here. The call creates or returns a singleton instance of the given class. The call in the `RandomInterface` initializer list looks like this

```
    _ri_manager(problem.getGlobal<RandomInterfaceManager>()),
```

### Rationale for the enhancement or information for reproducing the error

Empower developers to do stuff which they previously needed to modify `FEProblem` for, without having to modify `FEProblem`.

### Identified impact

Will allow us to move code out of `FEProblem`, easier maintenance, more developer flexibility.

P.S.: Yea, I know that it is totally possible to inherit from FEProblem. But if I need functionality in the phase_field module and create a `PhaseFieldFEProblem` and I need stuff in tensor_mechanics and create a `TensorMechanicsFEProblem` and now want to run a coupled simulation... I suddenly need to create yet another `FEProblem` derived class that combines functionality from both. 
With the system described above the needed functionality is hooked into `FEProblem` dynamically and the user does not have to jump through any hoops.

### PR DESCRIPTION
**Pull request by dschwen**
This adds the global singleton registration system to `FEProblem` (b4e897d) and demonstrates its use by moving `RandomInterface` related stuff out of `FEProblem` (5cecc1a). I foresee that we may need a few more hooks to this system (as in d9d40e5) to make the interface more powerful. 

I hope this will let us migrate the XFEM stuff over to this system as well (so that all code could just stay in the XFEM module). I also have a use in mind for phase field (and derivative parsed materials).

Closes #6777

### CODE REVIEW

**Comment by dschwen** [comment]
Hm, on my laptop at home I build with the C++11 shared pointer and it works. With the boost shared pointer implementation it fails, and I cannot make heads or tails of that error message :-/

**Comment by jwpeterson** [comment]
I agree that the error message is very confusing, but I think the problem is that you changed `_random_data` to a MooseSharedPointer:
```
MooseSharedPointer<RandomData> _random_data;
```
and for some reason, Boost shared_ptr cannot be initialized with NULL, i.e. the error is actually triggered by this line in the initialization list:
```
_random_data(NULL)
```

I think you can construct a Boost shared_ptr with a `nullptr`, but not NULL, as that is simply an integral constant and considered ambiguous.  As far as I can tell, this is a shortcoming of the Boost shared_ptr implementation that we will happily be getting away from when we switch to C++11.  For now just removing the `NULL` initializer will also fix the problem.

**Comment by dschwen** [comment]
Thanks @jwpeterson, now I see that the error message simply points to the _end_ of the initializer list. I was wondering what the error has to do with the quoted line.

File: framework/include/base/FEProblem.h [hunk 1]
```diff
@@ -1271,5 +1265,36 @@ FEProblem::finalizeUserObjects(const MooseObjectWarehouse<T> & warehouse)
   }
 }
 
+template <class T>
+T &
+FEProblem::getGlobal()
+{
+  // get the class name as map key
+  const std::string name = typeid(T).name();
```
Comment (permcody) [inline framework/include/base/FEProblem.h]: `typeid` is pretty consistent these days but we still use the demangle helper in most other places in the framework. We should probably do that here too for consistency. Just grep for "typeid" and you'll see the other uses of this throughout.

Comment (dschwen) [inline framework/include/base/FEProblem.h]: I know about demangle, but this is uses the type name as a map key only and does not report it to the user. All I need here is for `typeid(T).name()` to return a consistent result for successive calls with the same `T`.

**Pull request closed without merging**